### PR TITLE
Support more inner comments

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Babel-generator does not print inner comments in certain situations
| Patch: Bug Fix?          | Y
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR supports inner comments attached to generators, async methods and async do expressions. Because such inner comments must be block comments due to the noLineTerminator restrictions, we then ensure that such restrictions are respected by the babel-generator.

This PR also relaxes the condition when Babel prints single arrow binding `x => {}`. Before this PR we prints `(x) => {}` when `x` has attached comments. In this PR we relaxes the condition so that we can still output `/* leading */x/* trailing */ => {}`.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15008"></a>

### CODE REVIEW

**Review by nicolo-ribaudo — COMMENTED** [review]
Can you add this test? (I don't think we can properly print it and that's fine)

```js
class A {
  async /* 1 */ * /* 2 */ method /* 3 */ ( /* 4 */ ) /* 5 */ {}
}
```

**Comment by JLHwung** [comment]
Yeah. When an AST node has more than one empty spot, the inner comments are not sufficient for recreating code exactly from AST.

**Comment by liuxingbaoyu** [comment]
Can this be postponed for a while? I'm going to rebase this after the #14979 merge, it's a bit cumbersome to rebase that one. Thanks!

File: packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/arrow-single-param-type/output.js [hunk 1]
```diff
@@ -1 +1 @@
-const x = (foo /*: string*/) => {};
+const x = foo /*: string*/ => {};
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/arrow-single-param-type/output.js:1]: This is not a valid Flow comment. If the arrow function has a single parameter with a trailing block comment that starts with `:`, we should still print the parentheses.

Comment (JLHwung) [reply packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/arrow-single-param-type/output.js:1]: Sad that flow does not support that. In that case we can partially revert the arrow function commit.

Comment (liuxingbaoyu) [reply packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/arrow-single-param-type/output.js:1]: There doesn't seem to be a change here?

Comment (JLHwung) [reply packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/arrow-single-param-type/output.js]: Ah I forgot to update the plugin test fixtures.

**Review by liuxingbaoyu — APPROVED** [review]

**Pull request merged**